dace/frontend/python/newast.py

```diff
@@ -4188,9 +4215,26 @@ def visit_Call(self, node: ast.Call):
                     if mapping is not None:
                         # Two-step replacement (N -> __dacesym_N --> mapping[N])
                         # to avoid clashes
-                        symbolic.safe_replace(
-                            mapping,
-                            lambda m: sd.replace_properties_dict(newarr, m))
+                        for sym, symvalue in mapping.items():
+                            if str(sym) != str(symvalue):
+                                sd.replace_properties(
+                                    newarr, {
+                                        symbolic.symbol(str(sym)):
+                                        symbolic.symbol('__dacesym_' + str(sym))
+                                    }, sym, '__dacesym_' + sym)
+                        for sym, symvalue in mapping.items():
+                            if str(sym) != str(symvalue):
+                                sd.replace_properties(
+                                    newarr, {
+                                        symbolic.symbol('__dacesym_' + str(sym)):
+                                        symbolic.pystr_to_symbolic(symvalue)
+                                    }, '__dacesym_' + str(sym), symvalue)
```

Comment:
Why does this need to be done in two steps, and does it need to be over two passes over the mapping? Can you not consolidate this into a single loop?

Reply:
This is documented here: https://github.com/spcl/dace/blob/a654383d173881a981f34549c03cd90cc6db5594/dace/symbolic.py#L916

The code you commented on is not new btw, but it is the one that existed in master until yesterday. @tbennun refactored it to the `symbolic.safe_replace` method. However, as noted in the comment in the following lines, I had to revert the change because it doesn't work (yet) properly in the context of the `visit_Call` method.

Reply:
Imagine that you are updating the shape of an array given by `(5*M + N,)` and that the replacement dictionary is `{M: N, N: M}`, i.e., the shape must become `(5*N + M,)`. If you did it in a single loop with two replacements, then starting with `M` you get `5*M + N -> 5*__dace__sym_M + N -> 5* N + N`. Maybe I misunderstand what you mean.